flink/core/src/main/java/cz/o2/proxima/flink/core/CommitLogSourceFunction.java

```diff
@@ -0,0 +1,320 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.api.common.state.CheckpointListener;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.OperatorStateStore;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.runtime.state.JavaSerializer;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
+import org.apache.flink.streaming.api.watermark.Watermark;
+
+@Slf4j
+public class CommitLogSourceFunction<T> extends RichParallelSourceFunction<T>
+    implements CheckpointListener, CheckpointedFunction {
+
+  private static final String OFFSETS_STATE_NAME = "offsets";
+
+  private static int assignPartition(Partition partition, int numParallelSubtasks) {
+    return partition.getId() % numParallelSubtasks;
+  }
+
+  public static CommitLogSourceFunction<StreamElement> of(
+      RepositoryFactory repositoryFactory, List<AttributeDescriptor<?>> attributeDescriptors) {
+    return new CommitLogSourceFunction<>(
+        repositoryFactory, attributeDescriptors, new IdentityResultExtractor());
+  }
+
+  @FunctionalInterface
+  public interface ResultExtractor<T> extends Serializable {
+
+    T toResult(StreamElement element);
+  }
+
+  public static class IdentityResultExtractor implements ResultExtractor<StreamElement> {
+
+    @Override
+    public StreamElement toResult(StreamElement element) {
+      return element;
+    }
+  }
+
+  @SuppressWarnings("java:S1948")
+  private static class SourceLogObserver<T> implements LogObserver {
+
+    private final transient CountDownLatch completed = new CountDownLatch(1);
+    private final Set<Partition> seenPartitions = new HashSet<>();
+
+    private final SourceContext<T> sourceContext;
+    private final ResultExtractor<T> resultExtractor;
+
+    /**
+     * When restoring from checkpoint, we need to skip the first element in each partition, due to
+     * {@link ObserveHandle#getCurrentOffsets()} contract.
+     */
+    private final boolean skipFirstElementFromEachPartition;
+
+    private long watermark = Watermarks.MIN_WATERMARK;
+
+    @Nullable private volatile Throwable error = null;
+
+    private SourceLogObserver(
+        SourceContext<T> sourceContext,
+        ResultExtractor<T> resultExtractor,
+        boolean skipFirstElementFromEachPartition) {
+      this.sourceContext = sourceContext;
+      this.resultExtractor = resultExtractor;
+      this.skipFirstElementFromEachPartition = skipFirstElementFromEachPartition;
+    }
+
+    @Override
+    public boolean onError(Throwable error) {
```

Comment:
nit: it will be nice to somehow test also this situation. Maybe via passing an `ResultExtractor` which throws exception?

Reply:
👍 added a test case for this